certora/specs/SocialRecoveryModule.spec

```
@@ -1,26 +1,208 @@
+    function safeContract.isOwner(address owner) external returns (bool) envfree;
+
+    // Wildcard Functions (Because of use of ISafe interface in Social Recovery Module)
+    function _.isModuleEnabled(address module) external => safeIsModuleEnabled(calledContract, module) expect bool ALL; // `calledContract` is a special variable.
+    function _.isOwner(address owner) external => sumarizeSafeIsOwner(calledContract, owner) expect bool ALL;
+}
+
+// A summary function that asserts that all `ISafe.isModuleEnabled` calls are done
+// to the `safeContract`, returning the same result as `safeContract.isModuleEnabled(...)`.
+function safeIsModuleEnabled(address callee, address module) returns bool {
+    assert callee == safeContract;
+    return safeContract.isModuleEnabled(module);
 }
 
-// A setup function that requires Safe contract to enabled the Social Recovery
-// Module.
+// A summary function that asserts that all `ISafe.isOwner` calls are done
+// to the `safeContract`, returning the same result as `safeContract.isOwner(...)`.
+function sumarizeSafeIsOwner(address callee, address owner) returns bool {
+    assert callee == safeContract;
+    return safeContract.isOwner(owner);
+}
+
+// A setup function that requires Safe contract to enable the Social Recovery Module.
 function requireSocialRecoveryModuleEnabled() {
     require(safeContract.isModuleEnabled(currentContract));
 }
 
-// This is a dummy rule to verify the Safe Contract Setup with the Social Recovery
-// Module is working as intended.
-rule recoveryModuleCanBeDisabled {
-    env e;
-    address prevModule;
+// Setup function that `require`s the integrity of the guardians linked list
+// in the `GuardianStorage` contract. This is needed as `invariant`s can only
+// be proven for `currentContract` and not `guardianStorageContract`, however,
+// the integrity of this list is required for rules and invariants of the
+// recovery module. For proof of this integrity, see `GuardianStorage.spec`.
+function requireGuardiansLinkedListIntegrity(address guardian) {
+    uint256 index;
+    require index < guardianStorageContract.entries[safeContract].count;
+    require currentContract.isGuardian(safeContract, guardian) =>
+        guardianStorageContract.getGuardians(safeContract)[index] == guardian;
+    require guardianStorageContract.entries[safeContract].count == guardianStorageContract.countGuardians(safeContract);
+}
+
+// This integrity rule verifies that if the addGuardianWithThreshold(...) executes, then ensure that:
+// - the Social Recovery Module is enabled
+// - the caller to the Module has to be the Safe Contract
+// - the new guardian is added to the guardian list,
+// - and no other account (guardian or not) is affected.
+rule addGuardianWorksAsExpected(env e, address guardian, uint256 threshold, address otherAccount) {
+    uint256 currentGuardiansCount = guardianStorageContract.entries[safeContract].count;
+
+    requireGuardiansLinkedListIntegrity(guardian);
+
```

uber-style-nit: Might make sense to organize the requirements and bindings together:

```suggestion
    requireGuardiansLinkedListIntegrity(guardian);

    uint256 currentGuardiansCount = guardianStorageContract.entries[safeContract].count;
```

---

Done 👍🏾